File: scripts/health.js
#!/usr/bin/env node
// health.js — cluster health score report + confidence scoring
// Reads audit.log + .regret files + manifest to score cluster stability
// and compute confidence (HIGH/MEDIUM/LOW) per cluster.
//
// Usage:
//   node scripts/health.js
//   node scripts/health.js --sort fragile
//   node scripts/health.js --json

import { readFileSync, readdirSync, existsSync } from 'fs'
import { resolve, join, basename } from 'path'
import { computeConfidence, parseAuditForDrift } from './confidence.js'

const args      = process.argv.slice(2)
const sortBy    = args[args.indexOf('--sort') + 1] ?? 'health'
const jsonOutput = args.includes('--json')
const regretDir = resolve(process.cwd(), 'regrets')
const auditLog  = join(regretDir, 'audit.log')

// ─── Load manifest (for input counts) ──────────────────────────────────────────

let manifest = { clusters: [] }
const manifestPath = resolve(process.cwd(), 'regrets/manifest.json')
try {
  manifest = JSON.parse(readFileSync(manifestPath, 'utf8'))
} catch (e) {
  if (e.code === 'ENOENT') {
    console.error(`❌ regrets/manifest.json not found at ${manifestPath}. Run 'regret init' first.`)
    process.exit(1)
  }
  console.error(`❌ Invalid JSON in ${manifestPath}: ${e.message}. Fix the syntax and retry.`)
  process.exit(1)
}

// Build lookup: cluster-id -> input count
const inputCountMap = {}
for (const c of manifest.clusters || []) {
  inputCountMap[c.id] = (c.inputs || []).length
}

// ─── Parse audit.log ──────────────────────────────────────────────────────────

function parseAuditLog() {
  if (!existsSync(auditLog)) return {}
  const content = readFileSync(auditLog, 'utf8').trim()
  if (!content) return {}

  const events = {}
  const blocks = content.split('\n\n').filter(Boolean)

  for (const block of blocks) {
    const lines = block.trim().split('\n')
    const header = lines[0] // "2024-03-01T09:00:00Z  UPDATE  cluster-id"
    const parts  = header.trim().split(/\s+/)
    const type   = parts[1]
    const id     = parts[2]
    if (!id) continue
    if (!events[id]) events[id] = { updates: 0, drifts: 0, history: [] }
    if (type === 'UPDATE') events[id].updates++
    if (type === 'DRIFT')  events[id].drifts++
    events[id].history.push({ type, date: parts[0] })
  }

  return events
}

// ─── Parse .regret files ──────────────────────────────────────────────────────

function parseRegretMeta(content) {
  const meta = {}
  const metaSection = content.split('\n---\n')[0]
  for (const line of metaSection.split('\n')) {
    const colonIdx = line.indexOf(': ')
    if (colonIdx === -1) continue
    const key = line.slice(0, colonIdx)
    const val = line.slice(colonIdx + 2).trim()
    meta[key] = val
  }
  return meta
}

// ─── Health score (0–100) ─────────────────────────────────────────────────────

function scoreCluster({ updates, drifts, ageDays }) {
  let score = 100
  score -= updates * 15   // each intentional update = -15
  score -= drifts  * 25   // each drift = -25 (worse, indicates hidden bugs)
  if (ageDays < 3)  score -= 10  // brand new, not proven yet
  if (ageDays > 30) score += 5   // old and stable = bonus
  return Math.max(0, Math.min(100, score))
}

function healthLabel(score, { isNew = false } = {}) {
  if (isNew)       return { label: 'NEW',      bar: '░░░░░░', color: '🩵', note: '(newly captured — run drift to verify)' }
  if (score >= 90) return { label: 'SOLID',    bar: '██████', color: '✅' }
  if (score >= 70) return { label: 'GOOD',     bar: '█████░', color: '🟢' }
  if (score >= 50) return { label: 'UNSTABLE', bar: '███░░░', color: '🟡' }
  return              { label: 'FRAGILE',   bar: '██░░░░', color: '🔴' }
}

// ─── Main ─────────────────────────────────────────────────────────────────────

let regretFiles
try {
  regretFiles = readdirSync(regretDir).filter(f => f.endsWith('.regret'))
} catch {
  console.error(`❌ regrets/ directory not found. Run capture.js first.`)
  process.exit(1)
}

if (!regretFiles.length) {
  console.log(`No .regret files found. Nothing to report.`)
  process.exit(0)
}

const auditData = parseAuditLog()
const driftMap  = parseAuditForDrift(auditLog)
const now = Date.now()

const clusters = regretFiles.map(f => {
  const id      = basename(f, '.regret')
  const content = readFileSync(join(regretDir, f), 'utf8')
  const meta    = parseRegretMeta(content)
  const audit   = auditData[id] ?? { updates: 0, drifts: 0, history: [] }
  const captured = meta.captured ? new Date(meta.captured).getTime() : now
  const ageHours = (now - captured) / (1000 * 60 * 60)
  const ageDays  = Math.floor(ageHours / 24)
  const isNew    = ageHours < 72 && audit.updates === 0 && audit.drifts === 0
  const score    = scoreCluster({ updates: audit.updates, drifts: audit.drifts, ageDays })
  const health   = healthLabel(score, { isNew })

  // Confidence score
  const inputCount = inputCountMap[id] ?? 0
  const hasDriftOrUpdate = !!driftMap[id]
  const confidence = computeConfidence({ inputCount, ageDays, hasDriftOrUpdate })

  return { id, ageDays, score, health, isNew, confidence, inputCount, ...audit }
})

// Sort
const sorted = [...clusters].sort((a, b) => {
  if (sortBy === 'fragile') return a.score - b.score
  if (sortBy === 'age')     return b.ageDays - a.ageDays
  if (sortBy === 'confidence') return a.confidence.score - b.confidence.score
  return b.score - b.score  // default: healthiest first
})

// ─── Render ───────────────────────────────────────────────────────────────────

if (jsonOutput) {
  // JSON output mode
  const jsonResult = {
    clusters: sorted.map(c => ({
      id: c.id,
      fragility: Math.max(0, 100 - c.score) / 100,
      lastCapture: c.ageDays === 0 ? 'today' : `${c.ageDays}d ago`,
      score: c.score,
      label: c.health.label,
      isNew: c.isNew,
      updates: c.updates,
      drifts: c.drifts,
      confidence: c.confidence.label,
      confidenceScore: c.confidence.score,
    }))
  }
  console.log(JSON.stringify(jsonResult, null, 0))
} else {
  const COL = { id: 24, updates: 8, drifts: 7, age: 9, health: 16, confidence: 8, detail: 30 }

  console.log(`\nCLUSTER HEALTH REPORT`)
  console.log(`${'─'.repeat(90)}`)
  console.log(
    `${'cluster'.padEnd(COL.id)}` +
    `${'updates'.padEnd(COL.updates)}` +
    `${'drifts'.padEnd(COL.drifts)}` +
    `${'age'.padEnd(COL.age)}` +
    `${'health'.padEnd(COL.health)}` +
    `${'conf'.padEnd(COL.confidence)}` +
    `detail`
  )
  console.log(`${'─'.repeat(90)}`)

  for (const c of sorted) {
    const age = c.ageDays === 0 ? 'today' : `${c.ageDays}d`
    const healthStr = `${c.health.bar} ${c.health.label}`
    const confStr = c.confidence.label
    const inputStr = `${c.inputCount} input${c.inputCount !== 1 ? 's' : ''}`
    const detail = `${inputStr}, ${age} old`
    console.log(
      `${c.id.padEnd(COL.id)}` +
      `${String(c.updates).padEnd(COL.updates)}` +
      `${String(c.drifts).padEnd(COL.drifts)}` +
      `${age.padEnd(COL.age)}` +
      `${healthStr.padEnd(COL.health)}` +
      `${confStr.padEnd(COL.confidence)}` +
      `${detail}`
    )
  }

  console.log(`${'─'.repeat(90)}`)

  // ─── Legend ─────────────────────────────────────────────────────────────────

  console.log(`\nLegend:`)
  console.log(`  🩵 NEW      = recently captured, not yet verified`)
  console.log(`  ✅ SOLID    = stable, no changes detected`)
  console.log(`  🟢 GOOD     = minor changes, still healthy`)
  console.log(`  🟡 UNSTABLE = frequent changes, needs attention`)
  console.log(`  🔴 FRAGILE  = critical, high drift or update rate`)
  console.log(`\nConfidence: HIGH (>=0.8) | MEDIUM (>=0.5) | LOW (<0.5)`)
  console.log(`  Formula: F1(inputs)*0.5 + F2(age)*0.2 + F3(drift history)*0.3`)

  // ─── Recommendations ──────────────────────────────────────────────────────────

  const fragile  = sorted.filter(c => c.score < 50)
  const unstable = sorted.filter(c => c.score >= 50 && c.score < 70)
  const lowConf  = sorted.filter(c => c.confidence.label === 'LOW')

  if (fragile.length || unstable.length) {
    console.log(`\nRecommendations:`)
    for (const c of fragile) {
      if (c.updates >= 3) console.log(`  ${c.id.padEnd(COL.id)} → high update rate, consider splitting this cluster`)
      if (c.drifts  >= 1) console.log(`  ${c.id.padEnd(COL.id)} → drift detected, add normalize rules to manifest`)
    }
    for (const c of unstable) {
      console.log(`  ${c.id.padEnd(COL.id)} → monitor closely, ${c.updates} update(s), ${c.drifts} drift(s)`)
    }
  } else {
    console.log(`\n✅ All clusters are healthy. Safe to refactor.`)
  }

  if (lowConf.length) {
    console.log(`\nLow confidence clusters (add more inputs or wait for maturity):`)
    for (const c of lowConf) {
      console.log(`  ${c.id.padEnd(COL.id)} ${c.confidence.score}  (${c.inputCount} input${c.inputCount !== 1 ? 's' : ''}, ${c.ageDays}d old)`)
    }
  }

  const solid = sorted.filter(c => c.score >= 90 && !c.isNew)
  if (solid.length) {
    console.log(`\nDo not touch (SOLID contracts):`)
    solid.forEach(c => console.log(`  ${c.id}`))
  }

  console.log()
}

File: scripts/health.py
#!/usr/bin/env python3
# health.py — cluster health score report for Python + all stacks
# Reads audit.log + .regret files to score cluster stability
#
# Usage:
#   python scripts/health.py
#   python scripts/health.py --sort fragile

import sys
import os
import json
import re
from datetime import datetime, timezone

# ─── CLI args ─────────────────────────────────────────────────────────────────

def parse_args():
    args = sys.argv[1:]
    sort_by = 'health'

    i = 0
    while i < len(args):
        if args[i] == '--sort' and i + 1 < len(args):
            sort_by = args[i + 1]
            i += 2
        else:
            i += 1

    return sort_by


# ─── Parse audit.log ──────────────────────────────────────────────────────────

def parse_audit_log(audit_path):
    if not os.path.exists(audit_path):
        return {}

    try:
        with open(audit_path, 'r', encoding='utf-8') as f:
            content = f.read().strip()
    except FileNotFoundError:
        return {}

    if not content:
        return {}

    events = {}
    blocks = re.split(r'\n\n+', content)

    for block in blocks:
        lines = block.strip().split('\n')
        if not lines:
            continue

        header = lines[0].strip()
        parts = header.split()
        if len(parts) < 3:
            continue

        event_type = parts[1]
        cluster_id = parts[2]

        if cluster_id not in events:
            events[cluster_id] = {'updates': 0, 'drifts': 0, 'history': []}

        if event_type == 'UPDATE':
            events[cluster_id]['updates'] += 1
        if event_type == 'DRIFT':
            events[cluster_id]['drifts'] += 1

        events[cluster_id]['history'].append({
            'type': event_type,
            'date': parts[0],
        })

    return events


# ─── Parse .regret metadata ──────────────────────────────────────────────────

def parse_regret_meta(content):
    meta = {}
    meta_section = content.split('\n---\n')[0]
    for line in meta_section.split('\n'):
        colon_idx = line.find(': ')
        if colon_idx == -1:
            continue
        key = line[:colon_idx]
        val = line[colon_idx + 2:].strip()
        meta[key] = val
    return meta


# ─── Health score (0–100) ─────────────────────────────────────────────────────

def score_cluster(updates, drifts, age_days):
    score = 100
    score -= updates * 15   # each intentional update = -15
    score -= drifts * 25    # each drift = -25 (worse, indicates hidden bugs)
    if age_days < 3:
        score -= 10  # brand new, not proven yet
    if age_days > 30:
        score += 5   # old and stable = bonus
    return max(0, min(100, score))


def health_label(score, *, is_new=False):
    if is_new:
        return {'label': 'NEW',      'bar': '░░░░░░', 'color': '🩵', 'note': '(newly captured — run drift to verify)'}
    if score >= 90:
        return {'label': 'SOLID',    'bar': '██████', 'color': '✅'}
    if score >= 70:
        return {'label': 'GOOD',     'bar': '█████░', 'color': '🟢'}
    if score >= 50:
        return {'label': 'UNSTABLE', 'bar': '███░░░', 'color': '🟡'}
    return              {'label': 'FRAGILE',   'bar': '██░░░░', 'color': '🔴'}


# ─── Main ─────────────────────────────────────────────────────────────────────

def main():
    sort_by = parse_args()
    regret_dir = os.path.join(os.getcwd(), 'regrets')
    audit_log = os.path.join(regret_dir, 'audit.log')

    # Find .regret files
    try:
        regret_files = [f for f in os.listdir(regret_dir) if f.endswith('.regret')]
    except FileNotFoundError:
        print("❌ regrets/ directory not found. Run capture first.")
        sys.exit(1)

    if not regret_files:
        print("No .regret files found. Nothing to report.")
        sys.exit(0)

    audit_data = parse_audit_log(audit_log)
    now = datetime.now(timezone.utc).timestamp()

    clusters = []
    for f in regret_files:
        cluster_id = os.path.splitext(f)[0]
        with open(os.path.join(regret_dir, f), 'r', encoding='utf-8') as fh:
            meta = parse_regret_meta(fh.read())

        audit = audit_data.get(cluster_id, {'updates': 0, 'drifts': 0, 'history': []})

        captured_str = meta.get('captured', '')
        if captured_str:
            try:
                captured = datetime.fromisoformat(captured_str.replace('Z', '+00:00')).timestamp()
            except ValueError:
                captured = now
        else:
            captured = now

        age_hours = (now - captured) / 3600
        age_days = int(age_hours / 24)
        is_new = age_hours < 72 and audit['updates'] == 0 and audit['drifts'] == 0
        score = score_cluster(audit['updates'], audit['drifts'], age_days)
        health = health_label(score, is_new=is_new)

        clusters.append({
            'id': cluster_id,
            'ageDays': age_days,
            'score': score,
            'health': health,
            'isNew': is_new,
            'updates': audit['updates'],
            'drifts': audit['drifts'],
        })

    # Sort
    if sort_by == 'fragile':
        sorted_clusters = sorted(clusters, key=lambda c: c['score'])
    elif sort_by == 'age':
        sorted_clusters = sorted(clusters, key=lambda c: c['ageDays'], reverse=True)
    else:
        sorted_clusters = sorted(clusters, key=lambda c: c['score'], reverse=True)

    # ─── Render ───────────────────────────────────────────────────────────────

    COL = {'id': 30, 'updates': 8, 'drifts': 7, 'age': 9}

    print('\nCLUSTER HEALTH REPORT')
    print('─' * 72)
    print(
        f"{'cluster':<{COL['id']}}"
        f"{'updates':<{COL['updates']}}"
        f"{'drifts':<{COL['drifts']}}"
        f"{'age':<{COL['age']}}"
        'health'
    )
    print('─' * 72)

    for c in sorted_clusters:
        age = 'today' if c['ageDays'] == 0 else f"{c['ageDays']}d"
        note = f"  {c['health']['note']}" if c['health'].get('note') else ''
        print(
            f"{c['id']:<{COL['id']}}"
            f"{c['updates']:<{COL['updates']}}"
            f"{c['drifts']:<{COL['drifts']}}"
            f"{age:<{COL['age']}}"
            f"{c['health']['bar']} {c['health']['label']}{note}"
        )

    print('─' * 72)

    # ─── Legend ─────────────────────────────────────────────────────────────

    print('\nLegend:')
    print('  🩵 NEW      = recently captured, not yet verified')
    print('  ✅ SOLID    = stable, no changes detected')
    print('  🟢 GOOD     = minor changes, still healthy')
    print('  🟡 UNSTABLE = frequent changes, needs attention')
    print('  🔴 FRAGILE  = critical, high drift or update rate')

    # ─── Recommendations ──────────────────────────────────────────────────────

    fragile = [c for c in sorted_clusters if c['score'] < 50]
    unstable = [c for c in sorted_clusters if 50 <= c['score'] < 70]

    if fragile or unstable:
        print('\nRecommendations:')
        for c in fragile:
            if c['updates'] >= 3:
                print(f"  {c['id']:<{COL['id']}} → high update rate, consider splitting this cluster")
            if c['drifts'] >= 1:
                print(f"  {c['id']:<{COL['id']}} → drift detected, add normalize rules to manifest")
        for c in unstable:
            print(f"  {c['id']:<{COL['id']}} → monitor closely, {c['updates']} update(s), {c['drifts']} drift(s)")
    else:
        print('\n✅ All clusters are healthy. Safe to refactor.')

    solid = [c for c in sorted_clusters if c['score'] >= 90 and not c.get('isNew')]
    if solid:
        print('\nDo not touch (SOLID contracts):')
        for c in solid:
            print(f"  {c['id']}")

    print()


if __name__ == '__main__':
    main()
